In this notebook we will test the revision module 

In [1]:
## IMPORTs
from utils.task import Task
import json
import os
from src.pipeline.revision import revision
from dotenv import load_dotenv
from utils.prompt import load_prompt
import tiktoken
from tqdm import tqdm

load_dotenv()

True

In [2]:
# Function to load JSON data
def load_json_data(filepath):
    with open(filepath, 'r') as file:
        data = json.load(file)
    return data


# Function to create task object
def create_task(example):
    return Task(example)

In [3]:
# load the task data
filepath = "C:/Users\yousf\Bureau\ConvergenceAI\CHESS_Impl\data/test/subsampled_test.json"
data = load_json_data(filepath)
# load the retrieved entities
filepath_entities = "C:/Users\yousf\Bureau\ConvergenceAI\CHESS_Impl\data/test/retrieved_entities.json"
retrieved_entities = load_json_data(filepath_entities)
# load the retrieved context
filepath_context = "C:/Users\yousf\Bureau\ConvergenceAI\CHESS_Impl\data/test/retrieved_context.json"
retrieved_context = load_json_data(filepath_context)
# load the candidate SQL 
filepath_query = "C:/Users\yousf\Bureau\ConvergenceAI\CHESS_Impl\data/test/generated_candidate.json"
generated_candidate = load_json_data(filepath_query)

In [5]:
# test candidate generation module with a random sample
index = 1
example = data[index]
task = create_task(example)
db_path = os.getenv("DB_ROOT_PATH") + f"/{task.db_id}/{task.db_id}.sqlite"
model = ("gpt-4")
ans = revision(task=task, retrieved_entities=retrieved_entities[index],
               retrieved_context=retrieved_context[index], generated_candidate=generated_candidate[index],
               model=model,
               num_samples=1)
ans

You are an SQL expert and very smart data analyst.
Objective: Your primary goal is to ensure strict adherence to the database admin instructions and use the correct conditions while crafting an accurate SQL query. Failure to follow these instructions will result in an incorrect query.

Database Schema:
CREATE TABLE frpm
(
	CDSCode TEXT not null primary key,
	`Academic Year` TEXT null, --
	`County Code` TEXT null, -- examples: `02`
	`District Code` INTEGER null, --
	`School Code` TEXT null, -- description: School Code
	`County Name` TEXT null, -- examples: `Merced`
	`District Name` TEXT null, --
	`School Name` TEXT null, --
	`District Type` TEXT null, --
	`School Type` TEXT null, --
	`Educational Option Type` TEXT null, --
	`NSLP Provision Status` TEXT null, -- examples: `Provision 2`, `Lunch Provision 2`
	`Charter School (Y/N)` INTEGER null, --
	`Charter School Number` TEXT null, --
	`Charter Funding Type` TEXT null, --
	IRC INTEGER null, --
	`Low Grade` TEXT null, --
	`High Grade` TEX

{'chain_of_thought_reasoning': "To find the city location of the high school with specific criteria, I need to join the 'schools' and 'frpm' tables because the 'EILCode' and 'GSserved' are in 'schools', while 'NSLP Provision Status' and 'County Name' are in 'frpm'. The 'schools' table contains the 'City' column which is required for the output. The join condition is based on the 'CDSCode' which is a common key in both tables. The filters applied are for 'EILCode' to be 'HS' (High School), 'GSserved' to be '12' (grades 9-12), 'NSLP Provision Status' to be 'Lunch Provision 2', and 'County Name' to be 'Merced'.",
 'SQL': "SELECT T1.City FROM schools AS T1 JOIN frpm AS T2 ON T1.CDSCode = T2.CDSCode WHERE T1.EILCode = 'HS' AND T1.GSserved = '12' AND T2.`NSLP Provision Status` = 'Lunch Provision 2' AND T2.`County Name` = 'Merced'"}

In [6]:
#test revision module with the subsampled test
model = "gpt-4"
num_samples = 1
res = []
for index, sample in tqdm(enumerate(data)):
    task = create_task(sample)
    ans = revision(task=task, retrieved_entities=retrieved_entities[index],
                   retrieved_context=retrieved_context[index], generated_candidate=generated_candidate[index],
                   model=model,
                   num_samples=num_samples)
    res.append(ans)
    print(str(index) + " : ", ans)

1it [00:09,  9.75s/it]

0 :  {'chain_of_thought_reasoning': "To solve the query, I first identified the tables and columns relevant to the question. The question asks for atom IDs related to a specific molecule ('TR346') and the count of distinct bond types associated with this molecule. I used the 'atom' table to fetch atom IDs and joined it with the 'bond' table to access bond types. I filtered atoms by 'molecule_id' and ensured that 'bond_type' is not null to count distinct bond types. I grouped the results by 'atom_id' to ensure each atom ID is listed with its corresponding count of bond types.", 'SQL': "SELECT T1.atom_id, COUNT(DISTINCT T2.bond_type) AS bond_type_count FROM atom AS T1 JOIN bond AS T2 ON T1.molecule_id = T2.molecule_id WHERE T1.molecule_id = 'TR346' AND T2.bond_type IS NOT NULL GROUP BY T1.atom_id"}


2it [00:24, 12.47s/it]

1 :  {'chain_of_thought_reasoning': "To find the city location of the high school with specific criteria, I need to join the 'schools' and 'frpm' tables because the 'NSLP Provision Status' information is in the 'frpm' table and the school level and city information is in the 'schools' table. The conditions specified are that the school must be a high school ('EILCode' = 'HS'), located in Merced County, with 'NSLP Provision Status' as 'Lunch Provision 2', and the grade span served must be from 9 to 12. I will use an INNER JOIN to combine these tables on the 'CDSCode' field, which is a common key. The query will filter based on these conditions and select the 'City' from the 'schools' table.", 'SQL': "SELECT T1.City FROM schools AS T1 INNER JOIN frpm AS T2 ON T1.CDSCode = T2.CDSCode WHERE T1.EILCode = 'HS' AND T1.County = 'Merced' AND T2.`NSLP Provision Status` = 'Lunch Provision 2' AND T1.GSserved = '9-12'"}


3it [00:32, 10.44s/it]

2 :  {'chain_of_thought_reasoning': "To find the tallest hero published by Marvel Comics, I need to join the superhero table with the publisher table to filter superheroes based on the publisher name. I will use the superhero table to get the full name of the hero and the height to determine the tallest. The publisher table will help me filter by 'Marvel Comics'. I will use ORDER BY on height_cm in descending order and LIMIT 1 to get the tallest hero. I will ensure to filter out any superheroes with NULL height to adhere to the database admin instructions.", 'SQL': "SELECT T1.full_name FROM superhero T1 JOIN publisher T2 ON T1.publisher_id = T2.id WHERE T2.publisher_name = 'Marvel Comics' AND T1.height_cm IS NOT NULL ORDER BY T1.height_cm DESC LIMIT 1"}


4it [00:41, 10.10s/it]

3 :  {'chain_of_thought_reasoning': "To answer the question, I need to count the distinct users who are elders (Age > 65) and have obtained the 'Supporter' badge. I will join the 'users' table with the 'badges' table on the UserId to filter users who have the 'Supporter' badge. The condition for the badge name and the age will be applied in the WHERE clause. I will use COUNT(DISTINCT) to ensure each user is counted only once, even if they might have obtained the badge multiple times.", 'SQL': "SELECT COUNT(DISTINCT T1.Id) FROM users T1 JOIN badges T2 ON T1.Id = T2.UserId WHERE T1.Age > 65 AND T2.Name = 'Supporter'"}


5it [00:52, 10.40s/it]

4 :  {'chain_of_thought_reasoning': "The initial query attempted to count distinct users from New York who have both 'Teacher' and 'Supporter' badges. The error in the query was due to incorrect column referencing as 'T1.UserId' instead of 'T1.Id' since the 'users' table uses 'Id' as the primary key for users. The query needs to join the 'users' table with the 'badges' table twice to filter users who have both types of badges. The WHERE clause correctly filters by location and badge names. The revised query corrects the column name from 'T1.UserId' to 'T1.Id' to match the schema.", 'SQL': "SELECT COUNT(DISTINCT T1.Id) FROM Users T1 JOIN Badges T2 ON T1.Id = T2.UserId JOIN Badges T3 ON T1.Id = T3.UserId WHERE T1.Location = 'New York' AND T2.Name = 'Teacher' AND T3.Name = 'Supporter'"}


6it [01:03, 10.41s/it]

5 :  {'chain_of_thought_reasoning': "The question asks for the description about the ruling of the card named 'Condemn'. According to the database schema, the 'text' column in the 'rulings' table contains the description of the ruling. The 'rulings' table is linked to the 'cards' table through the 'uuid' column. Therefore, I need to join these tables to access the ruling text for the specific card. The card's name is stored in the 'name' column of the 'cards' table. I will use an INNER JOIN to connect these tables on the 'uuid' column and filter the results where the card's name is 'Condemn'.", 'SQL': "SELECT T2.text FROM cards AS T1 INNER JOIN rulings AS T2 ON T1.uuid = T2.uuid WHERE T1.name = 'Condemn'"}


7it [01:11,  9.72s/it]

6 :  {'chain_of_thought_reasoning': "To answer the question, I need to join the 'molecule' and 'bond' tables to count the number of double bonds for the molecule 'TR006' and determine if it is carcinogenic. The 'molecule_id' is used to join these tables, and the 'bond_type' column is filtered to count only double bonds (' = '). The 'label' column in the 'molecule' table indicates if the molecule is carcinogenic. I corrected the table names and ensured the WHERE clause correctly identifies the molecule 'TR006'.", 'SQL': "SELECT COUNT(*) AS double_bonds, T1.label FROM molecule AS T1 JOIN bond AS T2 ON T1.molecule_id = T2.molecule_id WHERE T1.molecule_id = 'TR006' AND T2.bond_type = '=' GROUP BY T1.label"}


8it [01:22, 10.03s/it]

7 :  {'chain_of_thought_reasoning': "To solve the query, I first identified the tables and columns relevant to the question. The question asks for the full name and college of a club member with the position of 'Secretary'. The 'member' table contains member details including position, and it links to the 'major' table through 'link_to_major', which in turn contains the 'college'. The 'position' column in the 'member' table directly corresponds to the position of the member. I joined the 'member' table with the 'major' table to access the college information. I ensured to filter out null values for the columns involved in the output and conditions, as per the admin instructions. I also used aliases for tables as required.", 'SQL': "SELECT T1.first_name, T1.last_name, T2.college FROM member AS T1 JOIN major AS T2 ON T1.link_to_major = T2.major_id WHERE T1.position = 'Secretary' AND T1.first_name IS NOT NULL AND T1.last_name IS NOT NULL AND T2.college IS NOT NULL"}


9it [01:29,  9.27s/it]

8 :  {'chain_of_thought_reasoning': "The initial query provided was incorrect because it referenced a non-existent table 'compounds'. The correct tables to use are 'bond' for bond types and 'molecule' for counting all molecules. The bond_type column in the bond table should be filtered for '#' to find triple bonds and should not be NULL to ensure accurate calculations. The percentage calculation involves counting entries with triple bonds and dividing by the total number of bonds, then multiplying by 100 to get a percentage. The query uses subqueries for these counts and ensures that the bond_type is not NULL in both subqueries for accurate results.", 'SQL': "SELECT (SELECT COUNT(*) FROM bond T1 WHERE T1.bond_type = '#' AND T1.bond_type IS NOT NULL) * 100.0 / (SELECT COUNT(*) FROM bond T2 WHERE T2.bond_type IS NOT NULL) AS percentage_of_triple_bonds"}


10it [01:46, 11.68s/it]

9 :  {'chain_of_thought_reasoning': "To answer the question about Francesco Parravicini's potential on a specific date, I joined the Player and Player_Attributes tables using the player_api_id as the common key. I filtered the results for the player name 'Francesco Parravicini' and the specific date '2010-08-30 00:00:00'. I also ensured that the potential column is not null to avoid any null results. The table names in the original query were incorrect ('players' instead of 'Player'), so I corrected these to match the schema provided.", 'SQL': "SELECT T2.potential FROM Player T1 INNER JOIN Player_Attributes T2 ON T1.player_api_id = T2.player_api_id WHERE T1.player_name = 'Francesco Parravicini' AND T2.date = '2010-08-30 00:00:00' AND T2.potential IS NOT NULL"}


11it [02:03, 13.28s/it]

10 :  {'chain_of_thought_reasoning': "To answer the question, I need to count the number of male clients living in the 'North Bohemia' region with an average salary greater than 8000. The 'client' table contains the gender and district_id, which links to the 'district' table where the region (A3) and average salary (A11) are stored. I will join these tables on district_id and apply the conditions on gender, A3, and A11. The COUNT function will be used to get the total number of such clients.", 'SQL': "SELECT COUNT(*) FROM client T1 JOIN district T2 ON T1.district_id = T2.district_id WHERE T1.gender = 'M' AND T2.A3 = 'North Bohemia' AND T2.A11 > 8000 AND T2.A11 IS NOT NULL"}


12it [02:13, 12.35s/it]

11 :  {'chain_of_thought_reasoning': "The initial query provided in the question was incorrect because it referenced non-existent columns such as 'name' and 'lowest_grade' in the Schools table. According to the schema, the correct column for the school's name is 'School', and for the lowest grade offered, it is 'GSoffered'. Additionally, the column for the city is 'City', and the state is specified in the 'State' column. The query needs to select the school name, city, and lowest grade from the Schools table where the state is 'CA', ensuring latitude is not null, and order the results by latitude in ascending order to find the lowest latitude. Only the first result is needed, hence the use of LIMIT 1.", 'SQL': "SELECT T1.School AS School_Name, T1.City, T1.GSoffered AS Lowest_Grade FROM Schools AS T1 WHERE T1.State = 'CA' AND T1.Latitude IS NOT NULL ORDER BY T1.Latitude ASC LIMIT 1"}


13it [02:26, 12.42s/it]

12 :  {'chain_of_thought_reasoning': "The query aims to calculate the percentage of bonds that are double bonds (' = ') for the molecule with ID 'TR047'. The calculation is done by dividing the number of double bonds by the total number of bonds for this molecule, then multiplying by 100 to get the percentage. The WHERE clause filters bonds belonging to 'TR047' and ensures that bond_type is not null to avoid incorrect calculations. The table name 'bonds' was incorrect in the initial query and has been corrected to 'bond'.", 'SQL': "SELECT (SUM(CASE WHEN bond_type = ' = ' THEN 1 ELSE 0 END) * 100.0 / COUNT(bond_id)) AS percent FROM bond T1 WHERE T1.molecule_id = 'TR047' AND T1.bond_type IS NOT NULL"}


14it [02:36, 11.68s/it]

13 :  {'chain_of_thought_reasoning': "To solve the query, I first identified the relevant tables: 'votes' for bounty information, 'posts' for post titles, and 'users' for user display names. I then joined these tables based on their relationships: 'votes' with 'posts' on PostId, and 'votes' with 'users' on UserId. The WHERE clause filters for a BountyAmount of 50 and titles containing 'variance'. The SELECT clause only includes the DisplayName from the 'users' table, as per the instructions.", 'SQL': "SELECT T3.DisplayName FROM votes T1 JOIN posts T2 ON T1.PostId = T2.Id JOIN users T3 ON T1.UserId = T3.Id WHERE T1.BountyAmount = 50 AND T2.Title LIKE '%variance%'"}


15it [02:45, 11.02s/it]

14 :  {'chain_of_thought_reasoning': "The initial query attempted to select a non-existent column 'DateObtained' from the 'Badges' table. The correct column name, as per the schema, is 'Date'. The table aliases were also incorrectly referenced. The 'Users' table should be aliased as T1 and the 'Badges' table as T2. The join condition should correctly reference 'T1.Id' and 'T2.UserId' based on the foreign key relationship defined in the schema. The WHERE clause correctly filters users based on their location in 'Rochester, NY'.", 'SQL': "SELECT T2.Date FROM Users T1 JOIN Badges T2 ON T1.Id = T2.UserId WHERE T1.Location = 'Rochester, NY'"}


In [7]:
from pathlib import Path

schema_path = Path('C:/Users/yousf/Bureau/ConvergenceAI/CHESS_Impl/data/test/final_sql.json')
with open(schema_path, 'w') as f:
    json.dump(res, f, indent=4)

## Cost Estimation per task

In [7]:
PROMPT_PATH = os.getenv("PROMPT_ROOT_PATH") + "\\candidate_generation.txt"
prompt = load_prompt(PROMPT_PATH)

In [8]:
def tokens_calc(example):
    encoding = tiktoken.get_encoding("cl100k_base")
    num_tokens = len(encoding.encode(example))
    return num_tokens

In [9]:
#prompt template tokens 
tokens_calc(prompt)

727

The prompt template has 727 tokens in total, but it have so many variables so it will more longer than
that to estimate the approximate token numbers we will see with a formatted prompt.
Let's say approximately 2200 tokens


In [10]:
## Output tokens estimation from an example 
output_example = """
{'chain_of_thought_reasoning': "To solve the query, I first identified the tables and columns relevant to the question. The question asks for the full name and college of a club member with the position of 'Secretary'. The 'member' table contains member details including position, and the 'major' table contains college information. The 'position' column in the 'member' table is used to filter for 'Secretary'. The 'link_to_major' in the 'member' table and 'major_id' in the 'major' table are used to join these tables. The 'college' column in the 'major' table provides the college information. I ensured to select only the columns explicitly asked for in the question, adhering to the database admin instructions.",
 'SQL': "SELECT T1.first_name, T1.last_name, T2.college FROM member AS T1 JOIN major AS T2 ON T1.link_to_major = T2.major_id WHERE T1.position = 'Secretary' AND T1.first_name IS NOT NULL AND T1.last_name IS NOT NULL AND T2.college IS NOT NULL"}
"""
tokens_calc(output_example)

228

Let's suppose the number of output tokens is 250

In [13]:
## Price calculation (just with gpt4 because in this module we don't use gpt3.5) 
input_price_per_token_gpt4 = 0.01 / 1000
output_price_per_token_gpt4 = 0.03 / 1000
price_gpt4 = 2200 * input_price_per_token_gpt4 + 250 * output_price_per_token_gpt4
print("estimated price per Task (GPT-4):", price_gpt4, "$")

estimated price per Task (GPT-4): 0.029500000000000002 $


In [14]:
## in this module there is a number of retrials so let's estimate the price with different number of retrials 

num_retrials = [1, 2, 3, 4, 5]
for num_retrial in num_retrials:
    total_price = price_gpt4 * num_retrial
    print("The estimated price with number of retrials of " + str(num_retrial) + ": ", total_price)
    

The estimated price with number of retrials of 1:  0.029500000000000002
The estimated price with number of retrials of 2:  0.059000000000000004
The estimated price with number of retrials of 3:  0.08850000000000001
The estimated price with number of retrials of 4:  0.11800000000000001
The estimated price with number of retrials of 5:  0.14750000000000002
